File: problems-extend-adapt/3801-3900/3805_rotation-twin-words/solution.go
func countRotationTwins(words []string) int64 {
	// Shifting a word by k adds k to every letter, so two words are
	// similar exactly when subtracting each word's own first letter
	// maps both onto the same normalized key: (c - word[0]) mod 26.
	counts := make(map[string]int64)
	for _, word := range words {
		base := int(word[0] - 'a')
		key := make([]byte, len(word))
		for i := 0; i < len(word); i++ {
			key[i] = byte('a' + (int(word[i])-'a'-base+26)%26)
		}
		counts[string(key)]++
	}
	// Pairs live inside one class; n <= 10^5 bounds the total by
	// n(n-1)/2 < 5 * 10^9, which fits int64 exactly.
	pairs := int64(0)
	for _, c := range counts {
		pairs += c * (c - 1) / 2
	}
	return pairs
}

File: problems-extend-adapt/3801-3900/3805_rotation-twin-words/solution.py
from typing import List


class Solution:
    def countRotationTwins(self, words: List[str]) -> int:
        # Shifting a word by k adds k to every letter, so two words are
        # similar exactly when subtracting each word's own first letter
        # maps both onto the same normalized key: (c - word[0]) mod 26.
        counts = {}
        for word in words:
            base = ord(word[0])
            key = "".join(chr((ord(c) - base) % 26 + ord("a")) for c in word)
            counts[key] = counts.get(key, 0) + 1
        # Pairs live inside one class; n <= 10^5 bounds the total by
        # n(n-1)/2 < 5 * 10^9, exact in Python's arbitrary-width ints.
        return sum(c * (c - 1) // 2 for c in counts.values())
